EduVPN-multiOS/Services/NotificationsService.swift

```diff
@@ -2,123 +2,200 @@
-            notificationContent.title = notification.title
-            if let body = notification.body {
-                notificationContent.body = body
+
+    static func isNotificationAllowed() -> Guarantee<Bool> {
+        return Guarantee<Bool> { callback in
+            notificationCenter.getNotificationSettings { settings in
+                callback((settings.authorizationStatus == .authorized) &&
+                    (settings.alertSetting == .enabled))
             }
-            let request = UNNotificationRequest(identifier: identifier, content: notificationContent, trigger: trigger)
-            
-            UNUserNotificationCenter.current().add(request, withCompletionHandler: callback)
-        } else {
-            let payload = NSUserNotification()
-            payload.title = notification.title
-            payload.informativeText = notification.body
-            payload.deliveryDate = NSCalendar.current.date(from: triggerDate)
-
-            NSUserNotificationCenter.default.scheduleNotification(payload)
-            
-            callback?(nil)
         }
     }
-    
-    #endif
-    
-    func permissionGranted(callback: @escaping (Bool) -> Void) {
-        #if os(iOS)
-        
-        UNUserNotificationCenter.current().getNotificationSettings { settings in
-            callback(settings.authorizationStatus == UNAuthorizationStatus.authorized)
+
+    static func scheduleCertificateExpiryNotification(for profile: Profile) -> Guarantee<Bool> {
+        guard let expiryDate = profile.api?.certificateModel?.x509Certificate?.notAfter else {
+            return Guarantee.value(false)
         }
-        
-        #elseif os(macOS)
-        
-        if #available(OSX 10.14, *) {
-            UNUserNotificationCenter.current().getNotificationSettings { settings in
-                callback(settings.authorizationStatus == UNAuthorizationStatus.authorized)
+
+        os_log("Certificate expires at %{public}@", log: Log.general, type: .debug, expiryDate as NSDate)
+
+        return requestAuthorization()
+            .then { isAuthorized -> Guarantee<Bool> in
+                guard isAuthorized else { return Guarantee.value(false) }
+                return isNotificationAllowed()
+            }.then { canAddNotificationRequest -> Guarantee<Bool> in
+                guard canAddNotificationRequest else { return Guarantee.value(false) }
+
+                let minutesToExpiry = Calendar.current.dateComponents([.minute], from: Date(), to: expiryDate).minute ?? 0
+                guard minutesToExpiry > 0 else { // Certificate has already expired
+                    return Guarantee.value(false)
+                }
+
+                // Normally, fire the notification 30 mins before expiry. If we're already past
+                // that time, fire it 2 seconds from now.
+                let maxMinutesFromNotificationToExpiry = 30
+                let minSecondsToNotification = 2
```

Comment:
For testing we might want to add something along the lines of...

```
                #if DEBUG
                let secondsToNotification = 10
                #else
                let secondsToNotification = (minutesToExpiry > maxMinutesFromNotificationToExpiry) ?
                    ((minutesToExpiry - maxMinutesFromNotificationToExpiry) * 60) : minSecondsToNotification
                precondition(secondsToNotification > 0)
                #endif
```

Reply:
I don't think I want that for every debug build (for example, if I were using a test server with a short expiry time, I don't want that). If I wanted to test a particular time, I'd change it in the code temporarily and test that.

Reply:
I can agree to that. :)